client.go

```diff
@@ -2907,3 +2902,75 @@ func releasePipelineWork(pool *sync.Pool, w *pipelineWork) {
 	w.err = nil
 	pool.Put(w)
 }
+
+// Resource Clean interface.
+// resourceClean can register into clientCleaner for cleaning resources.
+type resourceClean interface {
+	cleanResource()
+	hasResource() bool
+}
+
+// Cleaner to clean resources.
+// Client, HostClient register into cleaner. The Cleaner will check and delete client
+// where it has no resources.
+type clientCleaner struct {
+	initOnce       sync.Once
+	// clients to clean
+	clients        sync.Map
+}
+
+// Global cleaner.
+var cCleaner = &clientCleaner{}
+
+// Initialize cleaner.
+func (c *clientCleaner) init() {
+	c.initOnce.Do(func() {
```

Comment:
I think  `if atomic.CompareAndSwapUint32(&o.done, 0, 1) )` is better

clientCleaner  
```
type clientCleaner struct {
	done       int32
	// clients to clean
	clients        sync.Map
}
```
method init:
```
func (c *clientCleaner) init() {
	if atomic.CompareAndSwapUint32(&o.done, 0, 1) {
               go c.cleaner()
}
```

Reply:
Good catch, `sync.Once` is only really needed if what you are doing inside it is really required for the rest of the code to continue. This is not the case here, it's perfectly fine if the goroutine is started after calls to `register` have completed already.